inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/reduce_gpu_b_fs_yx_fsv16.cl

```diff
@@ -0,0 +1,406 @@
+    #elif REDUCE_L1_MODE
+        #if !INPUT0_IS_FP
+            acc += TO_ACCUMULATOR_TYPE(fabs(TO_FINAL_ACCUMULATOR_TYPE(input)));
+        #else
+            acc += fabs(input);
+        #endif
+    #elif REDUCE_LOG_SUM_EXP_MODE
+        #if !INPUT0_IS_FP
+            acc += TO_ACCUMULATOR_TYPE(exp(TO_FINAL_ACCUMULATOR_TYPE(input)));
+        #else
+            acc += exp(input);
+        #endif
+    #endif
+
+    return acc;
+}
+
+inline ACCUMULATOR_TYPE FUNC(sub_group_reduce)(ACCUMULATOR_TYPE acc) {
+    #if NEED_SUB_GROUP_REDUCE
+        #if REDUCE_SUM_MODE || REDUCE_MEAN_MODE || REDUCE_LOG_SUM_MODE
+            acc = sub_group_reduce_add(acc);
+        #elif REDUCE_MAX_MODE
+            acc = sub_group_reduce_max(acc);
+        #elif REDUCE_MIN_MODE
+            acc = sub_group_reduce_min(acc);
+        #elif REDUCE_PROD_MODE
+            ACCUMULATOR_TYPE next = ACCUMULATOR_VAL_ONE;
+            acc *= intel_sub_group_shuffle_down(acc, next, 8);
+            acc *= intel_sub_group_shuffle_down(acc, next, 4);
+            acc *= intel_sub_group_shuffle_down(acc, next, 2);
+            acc *= intel_sub_group_shuffle_down(acc, next, 1);
+            acc  = intel_sub_group_shuffle(acc, 0);
+        #elif REDUCE_AND_MODE
+            acc = sub_group_all(acc);
+        #elif REDUCE_OR_MODE
+            acc = sub_group_any(acc);
+        #elif REDUCE_SUM_SQUARE_MODE || REDUCE_L2_MODE
+            acc = sub_group_reduce_add(acc);
+        #elif REDUCE_L1_MODE
+            acc = sub_group_reduce_add(acc);
+        #elif REDUCE_LOG_SUM_EXP_MODE
+            acc = sub_group_reduce_add(acc);
+        #endif
+    #endif
+
+    return acc;
+}
+
+inline FINAL_ACCUMULATOR_TYPE FUNC(final_reduce)(FINAL_ACCUMULATOR_TYPE acc) {
+    #if REDUCE_MEAN_MODE
+        acc /= DIVIDER;
+    #elif REDUCE_L2_MODE
+        acc = sqrt(acc);
+    #elif REDUCE_LOG_SUM_MODE || REDUCE_LOG_SUM_EXP_MODE
+        acc = log(acc);
+    #endif
+
+    return acc;
+}
+
+inline uint FUNC(calc_linear_offset)(uint b, uint f, uint y, uint x) {
+    uint index = b * COMMON_OUTPUT_SIZE_X * COMMON_OUTPUT_SIZE_Y * COMMON_OUTPUT_FEATURE_NUM +
+                 f * COMMON_OUTPUT_SIZE_X * COMMON_OUTPUT_SIZE_Y +
+                 y * COMMON_OUTPUT_SIZE_X +
+                 x;
+
+    return index;
+}
+
+__attribute__((intel_reqd_sub_group_size(SIMD)))
+KERNEL(reduce_fsv16)(
+    const __global INPUT0_TYPE* data,
+    __global OUTPUT_TYPE* output
+#if HAS_FUSED_OPS_DECLS
+    , FUSED_OPS_DECLS
+#endif
+)
+{
+    const uint xy   = (uint)get_global_id(1) * READ_OFFSET;
+    const uint x    = xy % ((COMMON_OUTPUT_SIZE_X + READ_OFFSET - 1)/READ_OFFSET * READ_OFFSET);
+    const uint y    = xy / ((COMMON_OUTPUT_SIZE_X + READ_OFFSET - 1)/READ_OFFSET * READ_OFFSET);
+    const uint bf   = (uint)get_global_id(2) * SIMD;
+    const uint b    = bf / ((COMMON_OUTPUT_FEATURE_NUM + SIMD - 1)/SIMD * SIMD);
+    const uint f    = bf % ((COMMON_OUTPUT_FEATURE_NUM + SIMD - 1)/SIMD * SIMD);
```

Comment:
I think the following macros will be useful and will make the code more readable
```
#define CEIL_DIV(a, b) (((a) + (b) - 1)/(b))
#define ALIGN(a, b) (CEIL_DIV(a, b) * (b))
```